opal/mca/btl/sm/btl_sm_fbox.h

```diff
@@ -31,281 +52,276 @@
+#if OPAL_ENABLE_DEBUG
+            unsigned int old_end = ep->fbox_out.end;
+#endif
+            unsigned int remaining = buffer_free;
+
+            BTL_VERBOSE(("space needed for message: %" PRIsize_t ", remaining space in buffer: "
+                         "%u, checking for space at beginning of buffer",
+                         aligned_entry_size, remaining));
+
+            ep->fbox_out.end += remaining;
+            buffer_free = mca_btl_sm_fbox_out_free(&ep->fbox_out, fbox_size);
+            if (OPAL_UNLIKELY(buffer_free < aligned_entry_size)) {
+                /* not writing the skip token so give this space back */
+                ep->fbox_out.end -= remaining;
+                return NULL;
+            }
 
+            MCA_BTL_SM_FBOX_HDR(ep->fbox_out.buffer)->ival = 0;
+            opal_atomic_wmb();
+
+            BTL_VERBOSE(("writing a skip token at offset %u", old_end));
+            /* space is available. go ahead and mark remaining space to skip */
             mca_btl_sm_fbox_set_header(MCA_BTL_SM_FBOX_HDR(dst), 0xff, ep->fbox_out.seq++,
-                                       buffer_free - sizeof(mca_btl_sm_fbox_hdr_t));
-
-            end = MCA_BTL_SM_FBOX_ALIGNMENT;
-            /* toggle the high bit */
-            hbs = !hbs;
-            /* toggle the high bit match */
-            buffer_free = BUFFER_FREE(start, end, !hbm, fbox_size);
-            dst = ep->fbox_out.buffer + end;
+                                       remaining - sizeof(mca_btl_sm_fbox_hdr_t));
+            dst = ep->fbox_out.buffer;
         }
 
-        if (OPAL_UNLIKELY(buffer_free < size)) {
-            ep->fbox_out.end = (hbs << 31) | end;
-            opal_atomic_wmb();
-            OPAL_THREAD_UNLOCK(&ep->lock);
-            return false;
+        if (buffer_free < aligned_entry_size) {
+            return NULL;
         }
     }
 
-    BTL_VERBOSE(("writing fragment of size %u to offset %u {start: 0x%x, end: 0x%x (hbs: %d)} of "
+    BTL_VERBOSE(("writing fragment of size %u {start: 0x%x, end: 0x%x} of "
                  "peer's buffer. free = %u",
-                 (unsigned int) size, end, start, end, hbs, buffer_free));
+                 (unsigned int) aligned_entry_size, ep->fbox_out.start, ep->fbox_out.end, buffer_free));
+
+    ep->fbox_out.end += aligned_entry_size;
+    
+    /* zero-out the next */
+    if (buffer_free > aligned_entry_size) {
+        MCA_BTL_SM_FBOX_HDR(ep->fbox_out.buffer + (ep->fbox_out.end & fbox_offset_mask))->ival = 0;
+        opal_atomic_wmb();
+    }
+
+    return dst;
+}
+
+/* attempt to reserve a contiguous segment from the remote ep */
+static inline bool mca_btl_sm_fbox_sendi(mca_btl_base_endpoint_t *ep, unsigned char tag,
+                                         void *restrict header, const size_t header_size,
+                                         void *restrict payload, const size_t payload_size)
+{
+    size_t data_size = header_size + payload_size;
+
+    if (data_size > UINT_MAX) {
```

Comment:
This test is mostly unnecessary, as the immediate send shall only be used for very small data.

Reply:
I can remove the extra check. Will do that an request a re-review.

Reply:
Done.